File: Shadows_of_the_Knight_Episode_2/solution.py
import sys
import math

#
# Return a Matrix of booleans WxH
def BooleanMatrix(w,h):
#    m = [True * w] * h
    m = [[True for i in xrange(h)] for j in xrange(w)]
    return m

#
# Distance beetwen two points
def distance (p0,p1):
    x0,y0 = p0
    x1,y1 = p1

    return math.sqrt((x0-x1)*(x0-x1)+(y0-y1)*(y0-y1))

def centroid(m, w, h):
    k = 0
    x = 0
    y = 0
    for j in xrange(h):
        for i in xrange(w):
            if m[i][j]:
                x = x + i
                y = y + j
                k = k + 1
    p = (x//k, y//k)
    return p

def check_centroid(p,c,m,w,h):
    # Needs discard old visited point
    print >> sys.stderr, c, w, h
    x,y = c
    if c in p and m[x][y]:
        if x+1 <= w-1 and m[x+1][y] and not (x+1,y) in p:
            return (x+1,y)
        if x-1 >= 0 and m[x-1][y] and not (x-1,y) in p:
            return (x-1,y)
        if x+1 <= w-1 and y+1 <= h-1 and m[x+1][y+1] and not (x+1,y+1) in p:
            return (x+1,y+1)
        if x+1 <= w-1 and y-1 >= 0 and m[x+1][y-1] and not (x+1,y-1) in p:
            return (x+1,y-1)
        if y+1 <=h-1 and m[x][y+1] and not (x,y+1) in p:
            return (x,y+1)
        if y-1 >= 0 and m[x][y-1] and not (x,y-1) in p:
            return (x,y-1)
        if x-1 >= 0 and y + 1 <= h-1 and m[x-1][y+1] and not (x-1,y+1) in p:
            return (x-1,y+1)
        if x-1 >= 0 and y - 1 >= 0 and m[x-1][y-1] and not (x-1,y-1) in p:
            return (x-1,y-1)
    return c
#
# Discard Points
def discard_points (p0, p1, m, w, h, d):
    if cmp_distance(d) == 1:
        for j in xrange(h):
            for i in xrange(w):
                if distance((i,j),p1) > distance((i,j),p0):
                    m[i][j] = False
        return

    if cmp_distance(d) == -1:
        for j in xrange(h):
            for i in xrange(w):
                if distance((i,j),p1) < distance((i,j),p0):
                    m[i][j] = False
        return

    if cmp_distance(d) == 0:
        for j in xrange(h):
            for i in xrange(w):
                if distance((i,j),p1) != distance((i,j),p0):
                    m[i][j] = False
        return
    return


#
# Change String temperature for number: 1:Warmer, -1:Colder, 0:Same
def cmp_distance(d):
    if d == 'WARMER':
        return 1
    if d == 'COLDER':
        return -1
    # SAME
    return 0


# Auto-generated code below aims at helping you parse
# the standard input according to the problem statement.

# w: width of the building.
# h: height of the building.
w, h = [int(i) for i in raw_input().split()]
n = int(raw_input())  # maximum number of turns before game over.
x0, y0 = [int(i) for i in raw_input().split()]

M = BooleanMatrix(w,h)
P = []
# game loop
while True:
    bomb_dir = raw_input()  # Current distance to the bomb compared to previous distance (COLDER, WARMER, SAME or UNKNOWN)
    print >> sys.stderr,bomb_dir
    if bomb_dir == 'UNKNOWN':
        x1 = w/2
        y1 = h/2
#        x1, y1  = centroid(M,w,h)
    else:
        discard_points((x0,y0),(x1,y1),M,w,h,bomb_dir)
        x0 = x1
        y0 = y1
        x1, y1 = check_centroid(P,centroid(M,w,h),M,w,h)

    # Write an action using print
    # To debug: print >> sys.stderr, "Debug messages..."
    P.append((x1,y1))
    print >> sys.stderr, P
    print "%d %d" % (x1,y1)

File: Shadows_of_the_Knight_Episode_2/solution.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#define TRUE  0
#define FALSE 1
#define MAXSIZE   10000 
#define MOVEMENTS  100

struct point {
    int x;
    int y;
};


double distance(struct point p0, struct point p1)
{
    return sqrt( ((p0.x - p1.x)*(p0.x - p1.x)) + ((p0.y - p1.y)*(p0.y-p1.y)) ); 
}

struct point centroid(int mat[][MAXSIZE], int Xmin, int Xmax, int Ymin, int Ymax)
{
    int i,j;
    struct point p;
    unsigned long int X = 0;
    unsigned long int Y = 0;
    unsigned long int T = 0;
    for ( j = Ymin; j <= Ymax; j++ )
        for ( i = Xmin; i <= Xmax; i++ )
            if (mat[j][i] == TRUE)
            {
                X += (unsigned long int)i;
                Y += (unsigned long int)j;
                T++;
            }
    p.x = (int)(round(X / (double)T));
    p.y = (int)(round(Y / (double)T));
    return p;
}

void init_movements(struct point *p, int len)
{
    int i;
    for ( i = 0; i <= len-1; i++ )
        p[i].x = 0;
        p[i].y = 0;
}

void add_movement(struct point *p, struct point p0, int i, int len)
{
    if ( i >= len )
        return;
    p[i].x = p0.x;
    p[i].y = p0.y;
}

int check_movement(struct point *p, struct point p0, int len)
{
    int i;
    for (i=0; i <= len-1; i++)
        if (p[i].x == p0.x && p[i].y == p0.y)
            return 1;
    return 0;
}

struct point alternative_centroid(struct point C, struct point *p, int len,int mat[][MAXSIZE], int Xmax, int Ymax)
{
    struct point nC;
    int x, y;
    x = C.x;
    y = C.y;
    
    if ( x+1 <= Xmax && mat[y][x+1] == TRUE)
    {
        nC.x = x+1;
        nC.y = y;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( y+1 <= Ymax && mat[y+1][x] == TRUE)
    {
        nC.x = x;
        nC.y = y+1;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( x-1 >= 0 && mat[y][x-1] == TRUE)
    {
        nC.x = x-1;
        nC.y = y;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( y-1 >= 0 && mat[y-1][x] == TRUE)
    {
        nC.x = x;
        nC.y = y-1;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( x+1 <= Xmax && y-1 >= 0 && mat[y-1][x+1] == TRUE)
    {
        nC.x = x+1;
        nC.y = y-1;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( x-1 >= 0 && y+1 <= Ymax && mat[y+1][x-1] == TRUE )
    {
        nC.x = x-1;
        nC.y = y+1;
        if (!check_movement(p,nC,len))
            return nC;
    }
    if ( x+1 <= Xmax && y+1 <= Ymax && mat[y+1][x+1] == TRUE)
    {
        nC.x = x+1;
        nC.y = y+1;
        if (!check_movement(p,nC,len))
            return nC;    
    }    
    if ( x-1 >= 0 && y-1 >=0 && mat[y-1][x-1] == TRUE )
    {
        nC.x = x-1;
        nC.y = y-1;
        if (!check_movement(p,nC,len))
            return nC;
    }
}

int M[MAXSIZE][MAXSIZE] = {{0}};

/**
 * Auto-generated code below aims at helping you parse
 * the standard input according to the problem statement.
 **/
int main()
{
    struct point mov[MOVEMENTS];
    int Xmin;
    int Ymin;
    int Xmax;
    int Ymax;
    int i,j, k;
    struct point p0;
    struct point p1;
    struct point tmp;
    
    
    int W; // width of the building.
    int H; // height of the building.

    scanf("%d%d", &W, &H);
    int N; // maximum number of turns before game over.
    scanf("%d", &N);
    int X0;
    int Y0;
    int nmov = 0;
    scanf("%d%d", &X0, &Y0);

    init_movements(mov, MOVEMENTS);

    Xmin = 0;
    Ymin = 0;
    Xmax = W-1;
    Ymax = H-1;

    p0.x = X0;
    p0.y = Y0;
    p1.x = X0;
    p1.y = Y0;
    // game loop
    add_movement(mov,p0, nmov++, MOVEMENTS);
    
    while (1) {
        char bombDir[11]; // Current distance to the bomb compared to previous distance (COLDER, WARMER, SAME or UNKNOWN)
        scanf("%s", bombDir);
        if ((Xmax+1-Xmin) * (Ymax+1-Ymin) <= (1000*1000))
        {
        if (!strcmp(bombDir,"WARMER"))
        {
            for ( j = Ymin; j <= Ymax; j++ )
                for ( i = Xmin; i <= Xmax; i++ )
                {
                    tmp.x = i;
                    tmp.y = j;
                    if ( (M[j][i] == TRUE) && distance(p1,tmp) >= distance(p0,tmp) )
                        M[j][i] = FALSE;
                }
        }
        else if (!strcmp(bombDir,"COLDER"))
        {
            for ( j = Ymin; j <= Ymax; j++ )
                for ( i = Xmin; i <= Xmax; i++ )
                {
                    tmp.x = i;
                    tmp.y = j;
                    if ( (M[j][i] == TRUE) && distance(p1,tmp) <= distance(p0,tmp) )
                        M[j][i] = FALSE;
                }
        }
        else if (!strcmp(bombDir,"SAME"))
        {
            for ( j = Ymin; j <= Ymax; j++ )
                for ( i = Xmin; i <= Xmax; i++ )
                {
                    tmp.x = i;
                    tmp.y = j;
                    if ( (M[j][i] == TRUE) && distance(p1,tmp) != distance(p0,tmp) )
                        M[j][i] = FALSE;    
                }
                
        }
        // Write an action using printf(). DON'T FORGET THE TRAILING \n
        // To debug: fprintf(stderr, "Debug messages...\n");
        p0.x = p1.x;
        p0.y = p1.y;
        p1 =  centroid(M,Xmin,Xmax,Ymin,Ymax);
        if (check_movement(mov,p1,nmov))
            p1 = alternative_centroid(p1,mov,nmov,M,Xmax,Ymax);
        add_movement(mov,p1, nmov++, MOVEMENTS);
        }
        else
        {
            if (!strcmp(bombDir, "WARMER"))
            {
                if (p1.x > p0.x)
                    Xmin = p0.x + ( p1.x - p0.x );
                else if (p1.x < p0.x)
                    Xmax = p0.x - ( p0.x - p1.x );
                if (p1.y > p0.y)
                    Ymin = p0.y + (p1.y - p0.y);
                else if (p1.y < p0.y)
                    Ymax = p0.y - ( p0.y - p1.y );
            }
            if (!strcmp(bombDir, "COLDER"))
            {
                if (p1.x > p0.x)
                    Xmax = p0.x - ( p0.x - p1.x );
                else if (p1.x < p0.x)
                    Xmin = p0.x + ( p1.x - p0.x );

                if (p1.y > p0.y)
                    Ymax = p0.y - ( p0.y - p1.y );
                else if (p1.y < p0.y)
                    Ymin = p0.y + (p1.y - p0.y);
            }
            p0.x = p1.x;
            p0.y = p1.y;
            p1.x = (Xmax-Xmin)/2;
            p1.y = (Ymax-Ymin)/2;
            add_movement(mov,p1,nmov++, MOVEMENTS);
            fprintf(stderr,"%d,%d,%d,%d", Xmin,Xmax,Ymin,Ymax);
        }
        printf("%d %d\n", p1.x,p1.y);
    }

    return 0;
}
